### PR TITLE
Blocking command port when node is not caught up

### PR DESCRIPTION
**Pull request by danieldoglas**
### Details

This PR blocks the command port if a node is more than 12.5K commits behind the leader, and reopens when we're 1K behind.

### Fixed Issues
Fixes https://github.com/Expensify/Expensify/issues/437855

### Tests

_________
**Internal Testing Reminder:** when changing bedrock, please compile auth against your new changes

### CODE REVIEW

File: sqlitecluster/SQLiteNode.cpp [hunk 1]
```diff
@@ -1655,6 +1655,15 @@ void SQLiteNode::_onMESSAGE(SQLitePeer* peer, const SData& message) {
...
                     }
+                    const int64_t currentCommitDifference = message.calcU64("NewCount") - getCommitCount();
+                    const string blockReason = "COMMITS_LAGGING_BEHIND";
+                    if (currentCommitDifference > 50'000 && !_server.isCommandPortClosed(blockReason)) {
+                        SINFO("Node is lagging behind, closing command port so it can catch up.");
+                        _server.blockCommandPort(blockReason);
+                    } else if (currentCommitDifference < 10'000 && _server.isCommandPortClosed(blockReason)) {
+                        SINFO("Node is caught up enough, unblocking command port.");
+                        _server.unblockCommandPort(blockReason);
+                    }
```
Comment (tylerkaraszewski) [inline sqlitecluster/SQLiteNode.cpp:1666]: Is there a reason this block is here in particular? We have this data everywhere in this method, we don't need to do this only after starting a replicate thread (and it seems like we'd rather do it before starting a replicate thread, though I doubt it makes a practical difference).

File: sqlitecluster/SQLiteNode.cpp [hunk 2]
```diff
@@ -1655,6 +1655,15 @@ void SQLiteNode::_onMESSAGE(SQLitePeer* peer, const SData& message) {
                             }
                         }
                     }
+                    const int64_t currentCommitDifference = message.calcU64("NewCount") - getCommitCount();
+                    const string blockReason = "COMMITS_LAGGING_BEHIND";
+                    if (currentCommitDifference > 50'000 && !_server.isCommandPortClosed(blockReason)) {
```
Comment (tylerkaraszewski) [inline sqlitecluster/SQLiteNode.cpp:1660]: How did we pick 50,000?

**Review by tylerkaraszewski — CHANGES_REQUESTED** [review]

File: sqlitecluster/SQLiteNode.cpp [hunk 3]
```diff
@@ -1278,7 +1278,19 @@ void SQLiteNode::_onMESSAGE(SQLitePeer* peer, const SData& message) {
             peer->commandAddress = message["commandAddress"];
         }
         peer->setCommit(message.calcU64("CommitCount"), message["Hash"]);
-
+        
+        // We check the commit difference with 12,500 commits behind because that 
+        // represents ~30s of commits. If we're behind, let's close the command port
+        // so we can catch up with the cluster before processing new commands.
+        const int64_t currentCommitDifference = message.calcU64("NewCount") - getCommitCount();
```
Comment (tylerkaraszewski) [inline sqlitecluster/SQLiteNode.cpp:1285]: Not all messages have `newCount` but all messages have `CommitCount`, which we just checked a few lines above. But also we don't care what the peer's commit count is if it isn't leading.

Given the previous line, we can do:
```
if (peer == leadPeer && peer->commitCount > getCommitCount() + 12'500) 
```

And we don't need to parse the number from the message again.

**Review by tylerkaraszewski — CHANGES_REQUESTED** [review]

File: sqlitecluster/SQLiteNode.cpp [hunk 4]
```diff
@@ -1278,7 +1278,21 @@ void SQLiteNode::_onMESSAGE(SQLitePeer* peer, const SData& message) {
...
         peer->setCommit(message.calcU64("CommitCount"), message["Hash"]);
-
+        
+        // We check the commit difference with 12,500 commits behind because that 
+        // represents ~30s of commits. If we're behind, let's close the command port
+        // so we can catch up with the cluster before processing new commands.
+        const string blockReason = "COMMITS_LAGGING_BEHIND";
+        const int64_t currentCommitDifference =  peer->commitCount - getCommitCount();
+        if (peer == _leadPeer && currentCommitDifference >= 12'500 && !_blockedCommandPort) {
+            SINFO("Node is lagging behind, closing command port so it can catch up.");
```
Comment (cead22) [inline sqlitecluster/SQLiteNode.cpp:1288]: Let's log `currentCommitDifference`
```suggestion
            SINFO("Node is behind by " + SToStr(currentCommitDifference) + " commits, closing command port so it can catch up.");
```

File: sqlitecluster/SQLiteNode.cpp [hunk 5]
```diff
@@ -1278,7 +1278,21 @@ void SQLiteNode::_onMESSAGE(SQLitePeer* peer, const SData& message) {
...
+            _server.blockCommandPort(blockReason);
+            _blockedCommandPort = true;
+        } else if (currentCommitDifference < 1'000 && _blockedCommandPort) {
```
Comment (cead22) [inline sqlitecluster/SQLiteNode.cpp:1291]: The PR descriptions says 50K / 10K, but the diff shows 12.5K / 1K. Can we add a comment explaining the reasoning behind the 1K? The one for 12.5K is good because it says it's roughly 30s of commits

File: sqlitecluster/SQLiteNode.h [hunk 6]
```diff
@@ -364,9 +364,12 @@ class SQLiteNode : public STCPManager {
     // Stopwatch to track if we're giving up on the server preventing a standdown.
     SStopwatch _standDownTimeout;
 
-   // Our current State.
+    // Our current State.
     atomic<SQLiteNodeState> _state;
 
+    // Keeps track if we have closed the command port for commits fallen behind.
+    bool _blockedCommandPort{false};
```
Comment (cead22) [inline sqlitecluster/SQLiteNode.h:371]: Since we block the command port for other reasons, we should put the fact that this is being set based on the node being behind on the variable name, for clarity
```suggestion
    bool _blockedCommandPortForBeingBehind{false};
```

But do we really need this variable? Is there an easy way to check the reason the command port is blocked so we can do something like 
```cpp
_blockedCommandPortForBeingBehind = getReason() == "COMMITS_LAGGING_BEHIND";
if (peer == _leadPeer && currentCommitDifference >= 12'500 && !_blockedCommandPort) {
    ...
```

File: sqlitecluster/SQLiteNode.cpp [hunk 7]
```diff
@@ -1278,7 +1278,21 @@ void SQLiteNode::_onMESSAGE(SQLitePeer* peer, const SData& message) {
...
+            SINFO("Node is caught up enough, unblocking command port.");
```
Comment (cead22) [inline sqlitecluster/SQLiteNode.cpp:1292]: Let's log the commit difference and use closing/opening instead of closing/ublocking (we also have `SuppressCommandPort`, so let's not make this more confusing)
```suggestion
            SINFO("Node is caught up enough (behind by " + SToStr(currentCommitDifference) + " commits), re-opening command port.");
```

File: sqlitecluster/SQLiteNode.h (hunk 6 above)
Comment (danieldoglas) [reply sqlitecluster/SQLiteNode.h]: > But do we really need this variable? Is there an easy way to check the reason the command port is blocked so we can do something like

tl;dr: Adding this as a separate variable so we don't need to add complexity to the current methods related to block reasons. We had an implementation like you're suggesting before, but that would require adding new locks/changing current lock types.

This was discussed in this thread: https://expensify.slack.com/archives/CC7NECV4L/p1734981555759429?thread_ts=1733270912.693829&cid=CC7NECV4L

File: sqlitecluster/SQLiteNode.cpp (hunk 5 above)
Comment (danieldoglas) [reply sqlitecluster/SQLiteNode.cpp]: Changed the PR description, added comment.

**Review by tylerkaraszewski — APPROVED** [review]
